Praktikum/Tag02/uploads/01_strings_verbinden/strings_verbinden.c

```diff
@@ -2,23 +2,24 @@
 #include "strings_verbinden.h"
 
 void strings_verbinden(char *zusammen, char *string1, char *string2) {
```

Comment:
Ich habe gesehen, dass du diese Signatur von der Aufgabenstellung vorgegeben bekommen hast, darum ist das folgende nur als weiterführende Information dazu gedacht.

Diese Signatur ist aus Security-Sicht gefährlich. `string1` und `string2` sind potentiell user-controlled inputs. Das ist einer der wichtigsten Angriffswege: User-Eingaben. Dummerweise wird bei C-Strings die Länge bzw. das Ende durch das Nullbyte markiert. Wenn es ein Angreifer schafft den Buffer so voll zu schreiben, dass das Nullbyte am Ende fehlt, dann sind verschiedene Angriffe möglich. Leider klappt das häufiger als einem lieb ist, weil z.B. das Setzen des Nullbytes irgendwo anders im Code vergessen wird oder nur unter bestimmten Bedingungen passiert, die nicht immer zutreffen. Grundsätzlich die Idee das Ende des Strings in-band, d.h. mit den Daten zusammen statt als separate Meta-Daten zu führen, ist gefährlich, weil dann ein:e Angreifer:in potentiell die Möglichkeit hat das Nullbyte zu manipulieren.

Ganz konkret könnte ein Angriff hier so aussehen. Ich nehme an ich habe Kontrolle was in `string2` steht.
```
char string1[] = "Dosen't matter";
char string2[1] = {'a'}; # No Nullbyte!
char zusammen[65536];
strings_verbinden(zusammen, string1, string2);
```

Wenn das ausgeführt werden würde, dann würde der Inhalt des Speichers, genauer des Stacks, der hinter den Parametern liegt, ausgelesen werden. Wenn `zusammen` an den oder die Angreifer:in zurück geht, dann hat er oder sie jede Menge interne Informationen über das angegriffene System. Das ist ein sogenanntes Information Leak, sehr ähnlich der Heartbleed-Attacke: https://heartbleed.com/

Ein anderer konkreter Angriff wäre, wenn man irgendwie erreichen kann, dass `string1` und `string2` zusammen größer sind als `zusammen`. Dann könnte man den Stack überschreiben:
```
char string1[] = "Dosen't matter";
char string2[] = "Mein Angriffscode, der eigentlich aus total wilden Hexcodes bestehen würde";
char zusammen[20];
strings_verbinden(zusammen, string1, string2);
```

Hier würde alles hinter der Variablen `zusammen` auf dem Stack überschrieben werden. Das ist ein klassischer Buffer Overflow. Funktionsweise ist grob, dass man versucht den Return Pointer, der anzeigt wohin nach Beenden der aktuellen Funktion zurückgesprungen werden soll um das Programm weiter auszuführen, zu überschreiben. Der liegt im Stack hinter den übergebenen Parametern. Also muss ein:e Angreifer:in noch ein paar mehr Infos zusammen holen, um den Angriff durchzuführen, kann dann aber sagen welcher Code als nächstes ausgeführt werden soll. Das führt üblicherweise zur Code Execution und ist der Jackpott.

Diese Angriffe können verhindert werden, wenn der Code möglichst wenige Annahmen macht, sondern selbst alles prüft. Ist der Ausgabe-Buffer groß genug? Lese ich nicht mehr als der Input-String lang ist? In C ist das meist nur möglich indem die Länge eines Buffers separat übergeben wird. Also die Signatur z.B. so aussieht:
```
void strings_verbinden(char *zusammen, size_t len_zusammen, char *string1, size_t len_string1, char *string2, size_t len_string2)
```
Dann läuft man die Schleife solange durch bis entweder ein Nullbyte oder einer der `len_*` erreicht wurde. Und die `len_*` müssen natürlich von den Programmierern vorgegeben sein, statt den Angreifern.

Reply:
Oh, noch etwas. Viele C-Funktionen werden als insecure betrachtet. Beispielsweise `strcpy`, da sie ebenfalls eine solche Signatur haben. Es gibt bessere Alternativen wie z.B. `strncpy`, die auf ähnliche Weise angepasste Signaturen haben um Angriffe wie hier beschrieben zu erschweren.

Und falls es dir aufgefallen ist: `strncpy` bricht zwar ab, wenn der Input länger ist als der Output Buffer, schreibt in diesem Fall aber kein Nullbyte ans Ende des Output Buffers. Ein nicht-terminierter C-String ist eine Zeitbombe, die nur auf die richtige Funktion wartet. Deshalb gibt es auch noch `strlcpy`, das immer garantiert ein Nullbyte ans Ende zu schreiben.

Es ist m.E. ein super-wichtiger Skill für C-Programmierung auf Sicherheit zu achten. Es geht so schnell und einfach schief.